tardis/utilities/asyncbulkcall.py

```diff
@@ -0,0 +1,175 @@
+    before starting to execute them. The ``concurrent`` parameter controls whether
+    and how many bulks may run at once; when concurrency is low or disabled, tasks
+    may be waiting for execution even past ``size`` and ``delay``.
+    Possible values for ``concurrent`` are
+    :py:data:`False` (no concurrency, only one ``command`` at once),
+    either of :py:data:`True` or :py:data:`None` (unlimited concurrency),
+    or an integer above 0 to set a precise concurrency limit.
+
+    .. note::
+
+        If the ``command`` requires additional arguments,
+        wrap it via :py:func:`~functools.partial`.
+    """
+
+    def __init__(
+        self,
+        command: BulkCommand[T, R],
+        size: int,
+        delay: float,
+        concurrent: Union[int, bool, None] = True,
+    ):
+        self._command = command
+        self._size = size
+        self._delay = delay
+        self._concurrency = (
+            1
+            if concurrent is False
+            else sys.maxsize
+            if concurrent is None or concurrent is True
+            else concurrent
+        )
+        # task handling dispatch from queue to command execution
+        self._master_worker: Optional[asyncio.Task] = None
+        self._verify_settings()
+
+    @cached_property
+    def _concurrent(self) -> "asyncio.BoundedSemaphore":
+        """synchronized counter for active commands"""
+        return asyncio.BoundedSemaphore(value=self._concurrency)
+
+    @cached_property
+    def _queue(self) -> "asyncio.Queue[Tuple[T, asyncio.Future[R]]]":
+        """queue of outstanding tasks"""
+        return asyncio.Queue()
+
+    def _verify_settings(self):
+        if not isinstance(self._size, int) or self._size <= 0:
+            raise ValueError(f"expected 'size' > 0, got {self._size!r} instead")
+        if self._delay <= 0:
+            raise ValueError(f"expected 'delay' > 0, got {self._delay!r} instead")
+        if not isinstance(self._concurrency, int) or self._concurrency <= 0:
+            raise ValueError(
+                "'concurrent' must be one of True, False, None or an integer above 0"
+                f", got {self._concurrency!r} instead"
+            )
+
+    async def __call__(self, __task: T) -> R:
+        """Queue a ``task`` for bulk execution and return the result when available"""
+        result: "asyncio.Future[R]" = asyncio.get_event_loop().create_future()
+        await self._queue.put((__task, result))
+        self._ensure_worker()
+        return await result
+
+    def _ensure_worker(self):
+        """Ensure there is a worker to dispatch tasks for command execution"""
+        if self._master_worker is None:
+            self._master_worker = asyncio.ensure_future(self._bulk_dispatch())
+
+    async def _bulk_dispatch(self):
+        """Collect tasks into bulks and dispatch them for command execution"""
+        while True:
+            await asyncio.sleep(0)
```

Comment:
I've refactored this and ``_get_bulk`` for better task switching.

* Moved and commented the ``await asyncio.sleep(0)``. It will only trigger when the fast paths could be hit.
* The first item is fetched *before* starting the timeout. This allows to efficiently wait for a long time, instead of spin-waiting in ``delay`` intervals.